frontend/src/Chat.tsx

```diff
@@ -0,0 +1,54 @@
+import * as React from 'react';
+import classNames from 'classnames';
+
+export type Message = {from: string; message: string; from_game?: boolean};
+type Props = {
+  messages: Message[];
+};
+
+const Chat = (props: Props) => {
+  const [draft, setDraft] = React.useState<string>('');
+  const anchor = React.useRef(null);
+
+  // TODO: messages is mutable, so can't use reference equality to check for
+  // new state. After making messages immutable, update to only rely on
+  // props.messages
+  React.useEffect(() => {
+    anchor.current?.scrollIntoView({block: 'nearest', inline: 'start'});
+  }, [props.messages.length]);
```

Comment:
hm. Can we do a full array-equality check for `props.messages` here?

Not only is messages mutable, the onmessage handler bounds its length to 100 entries, so this would stop scrolling at some point?

Incidentally, how would we make messages immutable in this case?

Reply:
in the future, we can do something like this: https://github.com/rbtying/shengji/compare/danwang/state2?expand=1#diff-26b53db46abef847e920b00088370589R19

It computes a new array of messages and passes the new thing to the state to be stored.

When we rerender, we can use pointer equality to see whether or not it's different, and components will get the new copy passed down